### ISSUE TITLE
Service transaction changes tenantId when Tenants.withId is used to create an entry for a domain

### ISSUE THREAD

**Issue by jjballano** [grails/grails-data-mapping#1021]
### Steps to Reproduce
Having the following code, after removing some extra code don't needed here.

The domain:
```
class Payment implements MultiTenant<Payment> {
    BigDecimal amount
    static belongsTo = [user: User]
    Long tenantId
    def beforeInsert(){
        tenantId = user.id
        true
    }
}
```

and services:

```
@Transactional
class PaymentService {
    UserService userService 
    PaymentRepositoryService paymentRepositoryService

    void generateMonthlyPayments(){
        userService.findAll().each { user ->
           //users here will be different than 'admin' used in TenantResolver
            Tenants.withId(user.id){
                paymentRepositoryService.save(createPayment([user: user, amount: 0]))
            }
        }
    }

    private Payment createPayment(Map params){
        paymentRepositoryService.save(new Payment(params))
    }
}


@Service(Payment)
@CurrentTenant
@CompileStatic
class PaymentRepositoryService {
    
    Payment save(Payment payment) {
            payment?.save()
    }

}
```

And a TenantResolver:
```
//This is just for tests
class JobsTestTenantResolver implements TenantResolver{
    @Override
    Serializable resolveTenantIdentifier() throws TenantNotFoundException {
        return (Serializable) User.findByUsername('admin').id
    }
}
```

As I see, MultitenantEventListener is added/run twice (one after AutotimestampEventListener and another after HibernateEventListener), so first one changes tenantId to admin.id (``preInsertEvent.getEntityAccess().setProperty(tenantId.getName(), currentId)``), my beforeInsert changes it again to the user.id value but second time the event listener is executed, changes it again to admin.id and that what is committed to database.

### Expected Behaviour

It should creates an entry in the Payments table with tenantId same as user.id

### Actual Behaviour

TenantId is admin id.

### Environment Information

- **Operating System**: MacOS Sierra 10.12.6
- **GORM Version:** 6.1.8.RELEASE
- **Grails Version (if using Grails):** 3.3.1
- **JDK Version:** 1.8.0_60

**Comment by puneetbehl** [comment]
Please upload a sample application replicating the problem.

**Comment by jjballano** [comment]
Hi @puneetbehl,
Sure, you can find it here: https://github.com/jjballano/multitenancyError
You will see a failing test called MyServiceSpec

**Comment by puneetbehl** [comment]
I am not sure if I fully understand the problem because you are saving from `PaymentRepositoryService` which is annotated with `@CurrentTenant`. 

If you want to save against different tenant then use something like:
```
Tenants.withId(user.id) {
      Payment payment = paymentRepositoryService.save(createPayment([user: user, amount: 0]))
      payment.save()
 }
```

This way you don't need to override tenantId in `beforeInsert` or `beforeUpdate` methods.

**Comment by puneetbehl** [comment]
PR to your repo: https://github.com/jjballano/multitenancyError/pull/1

**Comment by puneetbehl** [comment]
Also, I think it's not a good practice to use `beforeInsert` or `beforeUpdate` to set `tenantId`.

**Comment by jjballano** [comment]
Sorry, I got rid of multitenancy in my project because of this problem some weeks ago and now I can't reproduce it again. It could be something else in my project as I had domains and services in a plugin which were use by a rest api (multitenant), a web project (not multitenant with an admin user logged) and jobs project (without user logged) and spring security plugin. Not easy to create a sample project.

So, close the issue if you want and I would reopen it again if I found the proper configuration that make this error shows up again.
